From ec73d35126ad9ab22192b1bbf0ac8a7d0fc95a60 Mon Sep 17 00:00:00 2001
From: Raphael Mounier <mounierr07@gmail.com>
Date: Sat, 15 Apr 2023 16:25:10 +0200
Subject: [PATCH 2/2] Huawei - fix all selinux denied rules on EMUI 9.1

Rules for
- Camera
- DRM widevine preavs
- Huawei debug
- oeminfo
- CPU gouvernor
---
 sepolicy/huawei.te         | 204 ++++++++++++++++++++++++++++++++++++-
 sepolicy/property_contexts |  10 +-
 2 files changed, 210 insertions(+), 4 deletions(-)

diff --git a/sepolicy/huawei.te b/sepolicy/huawei.te
index 2988ed6..077033a 100644
--- a/sepolicy/huawei.te
+++ b/sepolicy/huawei.te
@@ -6,14 +6,15 @@ type dmd_device, file_type;
 allow hal_fingerprint_server dmd_device:chr_file rw_file_perms;
 allow hal_fingerprint_server sysfs:file rw_file_perms;
 allow tee hal_fingerprint_default:process { getattr };
-allow tee teecd_data_file:dir { search read write create getattr add_name open };
-allow tee teecd_data_file:file { read write create getattr open };
-allow tee system_data_file:dir { getattr };
+allow tee teecd_data_file:dir { rw_dir_perms create setattr };
+allow tee teecd_data_file:file { rw_file_perms create setattr };
+allow tee system_data_file:dir { rw_dir_perms create setattr };
 
 type oeminfo_nvm, domain;
 type oeminfo_nvm_device, file_type;
 allowxperm oeminfo_nvm oeminfo_nvm_device:blk_file ioctl { 0x1260 };
 
+# offline charging
 allow charger rootfs:file { ioctl read getattr lock map execute entrypoint open };
 
 # This is use exclusively for init to relabel /dev/selinux from tmpfs to device
@@ -35,3 +36,200 @@ allowxperm vendor_init { teecd_data_file }:dir ioctl {
 type sys_block_mmcblk0, file_type;
 allow init sys_block_mmcblk0:lnk_file read;
 allow init sysfs_zram:lnk_file read;
+
+
+# --------------------- Not need to boot but this rules are denied  ----------------------------------------
+
+# Add support off splash2 Huawei logging
+type splash2_data_file, file_type;
+
+allow splash2_data_file self:filesystem { associate };
+allow init splash2_data_file:filesystem { mount relabelto relabelfrom };
+allow init splash2_data_file:dir { getattr mounton };
+allow vendor_init splash2_data_file:dir rw_dir_perms;
+allow vendor_init splash2_data_file:file rw_file_perms;
+allow vendor_init splash2_data_file:filesystem { getattr };
+allow installd splash2_data_file:filesystem { quotaget };
+allow rild splash2_data_file:dir rw_dir_perms;
+allow rild splash2_data_file:file rw_file_perms;
+allow kernel splash2_data_file:dir { rw_dir_perms create setattr rmdir };
+allow kernel splash2_data_file:file { rw_file_perms create unlink setattr };
+
+# Add support off modem Huawei logging
+type mnt_modem_file, file_type;  
+type modem_log_file, file_type;
+
+allow fsck mnt_modem_file:dir r_dir_perms;
+allow fsck modem_log_file:dir r_dir_perms;
+allow gmscore_app mnt_modem_file:dir r_dir_perms;
+
+# Fix CPU gouvernor (rules are in xml file)
+type hwemerffu_service, domain;
+allow hwemerffu_service proc:file r_file_perms;
+
+
+# Fix all SEdenied
+
+type hinetmanager, fs_type, sysfs_type;
+allow netutils_wrapper hinetmanager:fd { use };
+allow netutils_wrapper hinetmanager:fifo_file { write };
+allow hinetmanager self:capability { dac_override };
+
+allow mediaprovider_app cache_file:dir r_dir_perms;
+
+
+type hal_usb_default, domain;
+allow hal_usb_default self:capability { dac_override };
+
+type sysfs_led, domain;
+allow init sysfs_led:file { setattr };
+
+
+#allow crash_dump bluetooth_data_file:file rw_file_perms;
+#allow crash_dump resourcecache_data_file:file { read open };
+
+allow vold sys_block_mmcblk0:file { write };
+allow vold sysfs_zram:file { write };
+
+allow kernel device:dir rw_dir_perms;
+allow kernel self:capability { mknod fsetid };
+allow kernel device:chr_file { create setattr };
+
+type teecd_data_file_system, data_file_type;
+allow vendor_init device:chr_file { getattr };
+allow vendor_init teecd_data_file_system:dir { rw_dir_perms create setattr relabelto };
+allow vendor_init system_data_file:dir { rw_dir_perms create setattr };
+allow vendor_init system_data_file:lnk_file { create };
+
+allow init teecd_data_file:dir { mounton };
+allow init teecd_data_file:filesystem { mount relabelto relabelfrom };
+allow init sys_block_mmcblk0:file { open write };
+
+allow fsck block_device:blk_file { read open write ioctl };
+allow teecd_data_file self:filesystem { associate };
+allow rootfs labeledfs:filesystem { associate };
+allow tee self:capability { sys_ptrace };
+
+allow kernel device:chr_file { open write };
+allow oeminfo_nvm block_device:blk_file { ioctl getattr open read write };
+allow gpsdaemon system_data_file:lnk_file { read };
+allow oeminfo_nvm device:chr_file { ioctl open read write };
+allow rild system_data_file:lnk_file { read };
+allow kernel system_data_root_file:dir r_dir_perms;
+allow kernel system_data_root_file:file rw_file_perms;
+allow kernel unlabeled:file r_file_perms;
+
+allow hi110x_daemon rootfs:dir { open read };
+allow hi110x_daemon self:fifo_file { ioctl };
+
+
+allow hal_camera_default system_data_file:lnk_file { read };
+allow surfaceflinger bootanim:dir { search };
+allow surfaceflinger bootanim:file r_file_perms;
+allow hal_audio_default audioserver:fifo_file { write };
+allow installd self:capability { sys_ptrace };
+allow installd system_server:file { open read };
+allow installd system_server:dir { search };
+
+allow installd teecd_data_file:filesystem { quotaget };
+allow system_server sysfs_zram:lnk_file { read };
+allow system_server vendor_file:file r_file_perms;
+allow kernel sysfs_devices_system_cpu:file rw_file_perms;
+allow network_stack fs_bpf:file rw_file_perms;
+
+
+type hi110x_cust_data_file, file_type; 
+allow wpa_hisi hi110x_cust_data_file:dir r_dir_perms;
+
+
+allow kernel anr_data_file:dir r_dir_perms;
+allow bootanim system_data_file:dir r_dir_perms;
+
+allow priv_app gmscore_app:file { read };
+
+
+# OTPERASE = 0x4d19
+allowxperm oeminfo_nvm device:blk_file ioctl {
+  0x4d19
+};
+
+# TIOCGWINSZ = 0x5413
+allowxperm hi110x_daemon self:fifo_file ioctl {
+  TIOCGWINSZ
+};
+
+# BLKROSET = 0x125d 0x127c
+allowxperm phhsu_daemon device:blk_file ioctl {
+  BLKROSET
+  BLKDISCARDZEROES
+};
+
+allow shell self:netlink_kobject_uevent_socket { create setopt getopt bind read };
+allow init self:netlink_kobject_uevent_socket { create setopt getopt bind read };
+
+# DRM
+type hal_drm_widevine, domain;
+allow hal_drm_widevine shell_data_file:dir { search };
+allow hal_drm_widevine shell_test_data_file:dir { search };
+
+# AGPS-Supl 2.0 Huawei
+type hi110x_vendor_file, file_type;
+allow system_app hi110x_vendor_file:file r_file_perms;
+
+# For huawei EMUI 9 logging
+type dft_syshwdebug_prop, property_type;
+get_prop(rild, dft_syshwdebug_prop)
+get_prop(hal_nfc_default, dft_syshwdebug_prop)
+get_prop(hal_audio_default, dft_syshwdebug_prop)
+
+# Camera
+type cameradaemon_prop, property_type;
+get_prop(priv_app, cameradaemon_prop)
+
+
+# Authorize a lot of process to read default, config and system prop
+type hal_health_default, domain;
+type hal_camera_default, domain;
+type hal_audio_default, domain;
+type hal_keymaster_default, domain;
+type hal_nfc_default, domain;
+
+
+get_prop(system_app, cameradaemon_prop)
+
+get_prop(rild, config_prop)
+get_prop(rild, system_prop)
+get_prop(hal_health_default, default_prop)
+get_prop(hal_health_default, config_prop)
+get_prop(hal_fingerprint_default, default_prop)
+get_prop(hal_camera_default, default_prop)
+get_prop(hal_camera_default, config_prop)
+get_prop(hal_audio_default, default_prop)
+get_prop(hal_audio_default, config_prop)
+get_prop(hal_audio_default, system_prop)
+get_prop(hal_keymaster_default, config_prop)
+get_prop(hal_drm_widevine, default_prop)
+get_prop(hal_nfc_default, system_prop)
+
+type hi110x_daemon, domain;
+type wpa_hisi, domain;
+type aptouch_daemon, domain;
+type gpsdaemon, domain;
+type atcmdserver, domain;
+
+type hisecd, domain;
+type storage_info, domain;
+
+get_prop(hi110x_daemon, default_prop)
+get_prop(wpa_hisi, default_prop)
+get_prop(wpa_hisi, config_prop)
+get_prop(aptouch_daemon, system_prop)
+get_prop(gpsdaemon, default_prop)
+get_prop(oeminfo_nvm, default_prop)
+get_prop(atcmdserver, default_prop)
+get_prop(hisecd, default_prop)
+get_prop(mediaserver, vendor_default_prop)
+get_prop(storage_info, default_prop)
+get_prop(audioserver, vendor_default_prop)
+
+
diff --git a/sepolicy/property_contexts b/sepolicy/property_contexts
index 185a08b..2f22579 100644
--- a/sepolicy/property_contexts
+++ b/sepolicy/property_contexts
@@ -1,2 +1,10 @@
 persist.sys.sf.hs_mode         u:object_r:sf_hs_mode_prop:s0
-hwc.exynos.vsync_mode      u:object_r:vendor_hwc_prop:s0
+hwc.exynos.vsync_mode          u:object_r:vendor_hwc_prop:s0
+
+# GCAM
+vendor.camera.aux.packagelist  u:object_r:exported2_default_prop:s0 exact string
+persist.vendor.camera.privapp.list   u:object_r:exported2_default_prop:s0 exact string
+persist.vendor.camera.privapp.list2  u:object_r:exported2_default_prop:s0 exact string
+
+# Huawei debug vendor prop
+persist.sys.huawei.debug.on    u:object_r:dft_syshwdebug_prop:s0 exact int
-- 
2.25.1

